sled-agent/src/sled_agent.rs

```diff
@@ -942,3 +947,91 @@ impl SledAgent {
         self.inner.bootstore.clone()
     }
 }
+
+#[derive(From, thiserror::Error, Debug)]
+pub enum AddSledError {
+    #[error("Failed to learn bootstrap ip for {sled_id}")]
+    BootstrapAgentClient {
+        sled_id: Baseboard,
+        #[source]
+        err: bootstrap_agent_client::Error,
+    },
+    #[error("Failed to connect to DDM")]
+    DdmAdminClient(#[source] ddm_admin_client::DdmError),
+    #[error("Failed to learn bootstrap ip for {0}")]
+    NotFound(Baseboard),
+    #[error("Failed to initialize {sled_id}: {err}")]
+    BootstrapTcpClient {
+        sled_id: Baseboard,
+        err: crate::bootstrap::client::Error,
+    },
+}
+
+/// Add a sled to an  
+pub async fn add_sled_to_initialized_rack(
+    log: Logger,
+    sled_id: Baseboard,
+    request: StartSledAgentRequest,
+) -> Result<(), AddSledError> {
+    // Get all known bootstrap addresses via DDM
+    let ddm_admin_client = DdmAdminClient::localhost(&log)?;
+    let addrs = ddm_admin_client
+        .derive_bootstrap_addrs_from_prefixes(&[BootstrapInterface::GlobalZone])
+        .await?;
+
+    // Create a set of futures to concurrently map the baseboard to bootstrap ip
+    // for each sled
+    let mut addrs_to_sleds = addrs
+        .map(|ip| {
+            let log = log.clone();
+            async move {
+                let client = bootstrap_agent_client::Client::new(
+                    &format!("http://[{ip}]"),
+                    log,
+                );
+                let result = client.baseboard_get().await;
+
+                (ip, result)
+            }
+        })
+        .collect::<FuturesUnordered<_>>();
+
+    // Execute the futures until we find our matching sled or done searching
+    let mut target_ip = None;
+    while let Some((ip, result)) = addrs_to_sleds.next().await {
+        match result {
+            Ok(baseboard) => {
+                // Convert from progenitor type back to `sled-hardware`
+                // type.
+                let found = baseboard.into_inner().into();
+                if sled_id == found {
+                    target_ip = Some(ip);
+                    break;
+                }
+            }
+            Err(err) => {
+                warn!(
+                    log, "Failed to get baseboard for {ip}";
+                    "err" => #%err,
+                );
+            }
+        }
+    }
+
+    // Contact the sled and initialize it
+    let bootstrap_addr =
+        target_ip.ok_or_else(|| AddSledError::NotFound(sled_id.clone()))?;
+    let bootstrap_addr =
+        SocketAddrV6::new(bootstrap_addr, BOOTSTRAP_AGENT_RACK_INIT_PORT, 0, 0);
+    let client = crate::bootstrap::client::Client::new(
+        bootstrap_addr,
+        log.new(o!("BootstrapAgentClient" => bootstrap_addr.to_string())),
+    );
+
+    client.start_sled_agent(&request).await.map_err(|err| {
```

Comment:
What happens if we send this request to a sled that's already in the cluster?

Reply:
Great question. 

We'll end up in the [`ServerStarted` clause](https://github.com/oxidecomputer/omicron/blob/main/sled-agent/src/bootstrap/server.rs#L751) which tries to be idempotent but only on a few fields. It will currently return an error if the sled-id or address changed. 

I think it would probably be better to make this truly idempotent and check the whole request. What do you think?

Reply:
> I think it would probably be better to make this truly idempotent and check the whole request. What do you think?

It looks like as written, if we get a request with matching fields that are checked but non-matching other fields, we just ignore them entirely and return `Ok(_)`, which would presumably leave the client confused (it could reasonably assume all the params it passed are in use, but really the sled-agent was still running with some other set of params)? I agree making it check the whole request seems better.